src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -560,9 +709,16 @@ private void OnImportChangedAfterDelay(Task previous, object state)
             AssertIsForeground();
 
             var reference = (VisualStudioMetadataReference)state;
-            _changedReferencesPendingUpdate.Remove(reference);
+            ChangedReferencesPendingUpdate.Remove(reference);
+
+            bool hasMetadataReference;
+            lock (_gate)
+            {
+                hasMetadataReference = _metadataReferences.Contains(reference);
+            }
```

Comment:
I'd prefer if we just took the lock once in this function. Otherwise, nothing prevents this answer from changing from somebody else coming through as @jaredpar astutely observes. If/when we see performance problems, let's optimize this. Otherwise this is just asking for subtle races.